File: 정지민/[BOJ 1263] 시간 관리/1263.java
import java.util.*;
import java.io.*;

class Main {

    static class Task {
        int s;
        int t;

        public Task(int s, int t) {
            this.s = s;
            this.t = t;
        }
    }

    static int n;
    static Task[] tasks;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());
        tasks = new Task[n];

        for(int i=0; i<n; i++) {
            st = new StringTokenizer(br.readLine());
            int t = Integer.parseInt(st.nextToken());
            int s = Integer.parseInt(st.nextToken());
            tasks[i] = new Task(s, t);
        }

        Arrays.sort(tasks, (o1, o2) -> Integer.compare(o2.s, o1.s));

        int time = tasks[0].s;
        for(int i=0; i<n; i++) {
            int limit_time = tasks[i].s;
            int duration = tasks[i].t;

            if (limit_time >= time) {
                time -= duration;
            } else {
                time = limit_time - duration;
            }

            if (time < 0) {
                System.out.println(-1);
                return;
            }
        }

        System.out.println(time);
    }
}

File: 정지민/[BOJ 1263] 시간 관리/1263.py
'''
1. 아이디어: 마감기한이 늦은 것부터 채워간다.
    - 시간 순 그래프가 있다면 오른쪽(늦은 시간) 부터 채워가는 느낌으로 접근
    - 아이디어 떠올리는게 핵심..!!
    - 참고 블로그: https://nahwasa.com/entry/%EC%9E%90%EB%B0%94-%EB%B0%B1%EC%A4%80-1263-%EC%8B%9C%EA%B0%84-%EA%B4%80%EB%A6%AC-java
'''
n = int(input())
tasks = []
for _ in range(n):
    t, s = map(int, input().split())
    tasks.append((s, t))

tasks.sort(reverse = True)

time = tasks[0][0]

for i in range(n):
    limit_time = tasks[i][0]
    duration = tasks[i][1]

    if limit_time >= time:          # task 채워넣기
        time -= duration
    else:
        time = limit_time - duration
    
    if time < 0:
        print(-1)
        exit()

print(time)